homeassistant/components/husqvarna_automower/sensor.py

```diff
@@ -452,13 +453,34 @@ async def async_setup_entry(
                     for work_area_id in _work_areas
                     if description.exists_fn(_work_areas[work_area_id])
                 )
+                current_work_areas[mower_id] = set(_work_areas)
         entities.extend(
             AutomowerSensorEntity(mower_id, coordinator, description)
             for description in MOWER_SENSOR_TYPES
             if description.exists_fn(coordinator.data[mower_id])
         )
     async_add_entities(entities)
 
+    def _async_work_area_listener() -> None:
+        """Listen for new work areas and add sensor entities if they did not exist."""
+        for mower_id in coordinator.data:
+            if coordinator.data[mower_id].capabilities.work_areas:
+                _work_areas = coordinator.data[mower_id].work_areas
+                if _work_areas is not None:
+                    received_work_areas = set(_work_areas)
+                    new_work_areas = received_work_areas - current_work_areas[mower_id]
+                    if new_work_areas:
+                        current_work_areas[mower_id].update(new_work_areas)
+                        async_add_entities(
+                            WorkAreaSensorEntity(
+                                mower_id, coordinator, description, work_area_id
+                            )
+                            for description in WORK_AREA_SENSOR_TYPES
+                            for work_area_id in new_work_areas
+                        )
```

Comment:
Don't we remove the entities here as well?

Reply:
We remove all the work areas (number, sensor and switch) in the number platform:
https://github.com/home-assistant/core/blob/e8b6092948b6a7a22c311e6e228bad6e4db594b3/homeassistant/components/husqvarna_automower/number.py#L147-L156
It's not ideal, but it's easy and reliable, because we just look for the prefix `{mower_id}_{work_area_id}_` and delete all entities. Maybe we can move the function somewhere else?

Reply:
I would suggest moving that to `__init__.py` then. Since this code is not specific to one platform.